### PR TITLE
Initial Updates for issue 156 and intro

### PR DESCRIPTION
**Pull request by jsterne**

### CODE REVIEW

File: yang-module-versioning/draft-ietf-netmod-yang-module-versioning.xml [hunk 1]
```diff
@@ -434,6 +446,129 @@ revision 2019-01-02 {
...
+    may incorrectly reference a revision that is not in the history of the 
+    new revision. Further, the annotations might not apply depending on what 
+    the history is like in that new branch (e.g., an NBC change
+    that is ported might not be an NBC change in the new branch).
+    Schema node compatiblity extensions should not be copied over to the new 
+    branch. Instead, the changes should be 
+    considered as completely new on the new branch, and any compatibility information
+    should be generated from scratch.
+    </t>
+    <t>TBD: Is the format right in the example above ? quotes and semi colon ?</t>
```
Comment (xorrkaz) [inline yang-module-versioning/draft-ietf-netmod-yang-module-versioning.xml:536]: IMHO this is _allowed_, but we should also allow a child: description that explains why the change is one way or the other.  This is mostly obvious for some changes, but may be required a bit more in the case of a description-only semantic change.

File: yang-module-versioning/draft-ietf-netmod-yang-module-versioning.xml [hunk 2]
```diff
@@ -434,6 +446,129 @@ revision 2019-01-02 {
...
+    <t>TBD: Are we going to define when to add and not add these tags ? e.g.
+    optional anytime they are applicable, mandatory when a tool can't guess ?</t>
+    <t>
+    TBD: can the backwards-compatible-node tag be used in blatant contradiction to NBC rules ?  E.g. a range shrinks from 1..50 to 0..9. Is an author allowed to tag that leaf with the backwards-compatible-node tag ? Then what happens with the YANG Semver and the
```
Comment (xorrkaz) [inline yang-module-versioning/draft-ietf-netmod-yang-module-versioning.xml:540]: IMHO, no.  This is still NBC and this system gives one a chance to signal that.

File: yang-module-versioning/draft-ietf-netmod-yang-module-versioning.xml [hunk 3]
```diff
@@ -434,6 +446,129 @@ revision 2019-01-02 {
...
+    TBD: can the backwards-compatible-node tag be used in blatant contradiction to NBC rules ?  E.g. a range shrinks from 1..50 to 0..9. Is an author allowed to tag that leaf with the backwards-compatible-node tag ? Then what happens with the YANG Semver and the 
+    top level revision NBC tag ?
+    </t>
+    <t>
+    TBD: use the same non-backwards-compatible tag as the top level, but allow it to be applied to nodes ?
+    </t>
+    <t>
+    TBD: The scheme comparison draft limited the use of per-node tags to specific
+    YANG statements: description, must and when (and probably 'pattern' should
+    have been in that list). Are we going to limit their applicability ?
```
Comment (xorrkaz) [inline yang-module-versioning/draft-ietf-netmod-yang-module-versioning.xml:549]: It may be that we broaden this to address WG comments, but recommend judicious use to those situations where the NBC/BC change is not so terribly obvious.

File: yang-module-versioning/draft-ietf-netmod-yang-module-versioning.xml [hunk 4]
```diff
@@ -434,6 +446,129 @@ revision 2019-01-02 {
...
+    </t>
+    <t>
+    TBD: Change the name of the module level annotation from rev:non-backwards-compatible to
```
Comment (xorrkaz) [inline yang-module-versioning/draft-ietf-netmod-yang-module-versioning.xml:552]: I'm good with that.

File: yang-module-versioning/draft-ietf-netmod-yang-module-versioning.xml [hunk 5]
```diff
@@ -434,6 +446,129 @@ revision 2019-01-02 {
...
+    rev:non-backwards-compatible-revision (or rev:revision-non-backwards-compatible), to
+    distinguish more clearly from the per node annotation ?
+    </t>
+    <t>TBD: Allow the use of per node tags in pre-release versions ? If so,
```
Comment (xorrkaz) [inline yang-module-versioning/draft-ietf-netmod-yang-module-versioning.xml:556]: Eh, I think we say not to do this.  Keeps pre-release things a bit simpler.

File: yang-module-versioning/draft-ietf-netmod-yang-module-versioning.xml (hunk 1 above)
Comment (jsterne) [reply yang-module-versioning/draft-ietf-netmod-yang-module-versioning.xml:536]: Maybe we want this:
   rev:nbc-change-at 3.0.0 {
       rev:description "renamed the leaf from foo-1";  <- optional
   }

File: yang-module-versioning/draft-ietf-netmod-yang-module-versioning.xml (hunk 4 above)
Comment (jsterne) [reply yang-module-versioning/draft-ietf-netmod-yang-module-versioning.xml:552]: In the call today: Rob inclined to leave the name alone, Jason inclined to add "-revision" to the end. Everyone else abstained :-)

File: yang-module-versioning/draft-ietf-netmod-yang-module-versioning.xml (hunk 3 above)
Comment (jsterne) [reply yang-module-versioning/draft-ietf-netmod-yang-module-versioning.xml:549]: The per-node tags are really per-YANG-statement tags but for a subset of YANG statements (probably "data nodes" + rpc, notification, action, etc).
But we do probably want these tags to apply to choice, case, and grouping.

File: yang-module-versioning/draft-ietf-netmod-yang-module-versioning.xml [hunk 6]
```diff
@@ -434,6 +446,129 @@ revision 2019-01-02 {
...
+    new revision. Further, the annotations might not apply depending on what 
+    the history is like in that new branch (e.g., an NBC change
+    that is ported might not be an NBC change in the new branch).
+    Schema node compatiblity extensions should not be copied over to the new 
+    branch. Instead, the changes should be 
+    considered as completely new on the new branch, and any compatibility information
+    should be generated from scratch.
+    </t>
+    <t>TBD: Is the format right in the example above ? quotes and semi colon ?</t>
+    <t>TBD: Are we going to define when to add and not add these tags ? e.g.
```
Comment (jsterne) [inline yang-module-versioning/draft-ietf-netmod-yang-module-versioning.xml:537]: We didn't finish debating this on the call today, but one thing we thought we should add:
Nothing about compatibility of a node/statement can be inferred from teh absence of the granular compatibility statements.

File: yang-module-versioning/draft-ietf-netmod-yang-module-versioning.xml [hunk 7]
```diff
@@ -536,37 +536,55 @@ revision 2019-01-02 {
...
-	</list>	
+      <t>When a tool may not be able to accurately determine the compatibilty of a change. For example, a change in a 'pattern' or 'must' statement can be difficult for a user or tool to determine if it is a compatible change.</t>
+      <t>When a pattern, range or other statement is changed to more correctly define the server constraint. An example is correcting a pattern that incorrectly included 355.xxx.xxx.xxx as a possible IPv4 address to make it only accept up to 255.xxx.xxx.xxx.</t>
+    </list> 
     <t>Nothing about the compatibilty of a YANG statement can be inferred
-    from the absence of a compatibility extension statement. A leaf, for example, 
-    could be changed from a string type to a uint8 type.
-    Even if there is no rev:nbc-change-at substatement against that leaf,
-    the leaf is still considered as having a non-backwards-compatible change.
+    from the absence of a compatibility extension statement.
```
Comment (jsterne) [inline yang-module-versioning/draft-ietf-netmod-yang-module-versioning.xml:550]: updated proposal:    <t>Nothing about the backwards compatibility of a YANG statement is implied by the absence of a compatibility extension statement.  Hence, the YANG statement must be compared between the two revisions to determine the backwards compatibility.

File: yang-module-versioning/draft-ietf-netmod-yang-module-versioning.xml (hunk 1 above)
Comment (jsterne) [reply yang-module-versioning/draft-ietf-netmod-yang-module-versioning.xml]: description added to the PR

File: yang-module-versioning/draft-ietf-netmod-yang-module-versioning.xml (hunk 2 above)
Comment (jsterne) [reply yang-module-versioning/draft-ietf-netmod-yang-module-versioning.xml:591]: We ended up allowing this in the end. Updated in the PR.

File: yang-module-versioning/draft-ietf-netmod-yang-module-versioning.xml (hunk 6 above)
Comment (jsterne) [reply yang-module-versioning/draft-ietf-netmod-yang-module-versioning.xml:588]: updated in the PR

File: yang-module-versioning/draft-ietf-netmod-yang-module-versioning.xml [hunk 8]
```diff
@@ -525,84 +528,64 @@ revision 2019-01-02 {
...
-    <t>YAMG statement compatibility statements are not supported on 
+    <t>A node compatibility statement at a leaf or leaf-list context
+    takes precedence over a node compatibility statement in a typedef
+    used by the leaf or leaf-list. If multiple typedefs with 
+    compatibility statements are used by a leaf or leaf-list (e.g. a union), and there 
+    is no compatibility statement at the leaf or leaf-list context, then
+    the order of precedence used to classify the compatibility of the leaf
+    or leaf-list is as follows: nbc-change-at, bc-change-at and finally editorial-change-at. That is, the leaf or leaf-list takes the most impactful change 
+    classification of all the typedefs.</t>
+    <t>Node compatibility statements are not supported on YANG
```
Comment (jsterne) [inline yang-module-versioning/draft-ietf-netmod-yang-module-versioning.xml:578]: For the last part, focus on the union construct since that can also be used within a typedef. 
use significant instead of impactful

File: yang-module-versioning/draft-ietf-netmod-yang-module-versioning.xml [hunk 9]
```diff
@@ -525,84 +528,64 @@ revision 2019-01-02 {
...
+    revision matches the argument of any nbc-change-at statements. If any revision statements are removed, then all node compatibiilty statements that reference that revision MUST also be removed. Conversely, node compatibilty statements MUST NOT be removed unless the
     associated revision statement in the revision history is removed.
     </t>
-    <t>If a compatiblity extension is added to a 'grouping' statement,
+    <t>If a node compatiblity statement is added to a grouping,
     then all instances where the grouping is used in the module or
-    by an importing modules are also impacted by the compatibilty information. 
+    by an importing module are also impacted by the compatibilty information. 
     Similarly for a 'typedef', all 
     leafs and list-lists that use that typedef share the specified
```
Comment (jsterne) [inline yang-module-versioning/draft-ietf-netmod-yang-module-versioning.xml:563]: Add typedef into the leaf/leaf-list items

File: yang-module-versioning/draft-ietf-netmod-yang-module-versioning.xml [hunk 10]
```diff
@@ -447,17 +451,16 @@ revision 2019-01-02 {
     </figure>
   </section>
   
-  <section anchor="node-tags" title="Schema node compatibility extension statements">
-    <t>In addition to the revision extension statement in <xref target="revision_nbc_marker"/>, YANG extension statements are also available to indicate compatibility information
-    for certain individual YANG statements.
+  <section anchor="node-tags" title="Node compatibility extension statements">
+    <t>In addition to the revision extension statement in <xref target="revision_nbc_marker"/>, YANG extension statements are also available to indicate compatibility information for schema nodes and certain individual YANG statements.
     </t>
```
Comment (jsterne) [inline yang-module-versioning/draft-ietf-netmod-yang-module-versioning.xml:456]: Add a paragraph to say schema node as per 7950 and the statements listed in the YANG module defiition of nbc-change-at (forward ref).

File: yang-module-versioning/draft-ietf-netmod-yang-module-versioning.xml [hunk 11]
```diff
@@ -136,10 +136,12 @@
...
         ordering or backwards compatibility between different revisions of the same module.</t>
-    <t>Backwards-compatible (BC) change: A backwards-compatible change between two YANG module revisions, as defined
-    in <xref target="bc_update_rules"/></t>
+        <t>Backwards-compatible (BC) change: A backwards-compatible change between two YANG module revisions, as defined
+        in <xref target="bc_update_rules"/></t>
         <t>Non-backwards-compatible (NBC) change: A non-backwards-compatible change between two YANG module revisions, as
         defined in <xref target="nbc_update_rules"/></t>
+		<t>Node compatibility statement: An extension statements (e.g. nbc-change-at)
+		that can be used to indicate the backwards compatibility of individual schema nodes and specific YANG statements.</t>
         </list>
```
[the remark this answers is not in the document]
Comment (jsterne) [reply yang-module-versioning/draft-ietf-netmod-yang-module-versioning.xml:145]: done in local draft - will be part of next checkin in this PR.

File: yang-module-versioning/draft-ietf-netmod-yang-module-versioning.xml (hunk 10 above)
Comment (jsterne) [reply yang-module-versioning/draft-ietf-netmod-yang-module-versioning.xml:456]: done in local draft - will be part of next checkin in this PR.

File: yang-module-versioning/draft-ietf-netmod-yang-module-versioning.xml (hunk 9 above)
Comment (jsterne) [reply yang-module-versioning/draft-ietf-netmod-yang-module-versioning.xml:563]: done in local draft - will be part of next checkin in this PR.

File: yang-module-versioning/draft-ietf-netmod-yang-module-versioning.xml [hunk 12]
```diff
@@ -201,6 +203,8 @@
         <t>A statement that is defined using the YANG "extension" statement MAY be added, removed, or changed,
         if it does not change the semantics of the module.  Extension statement definitions SHOULD specify whether
         adding, removing, or changing statements defined by that extension are backwards-compatible or non-backwards-compatible.</t>
+        <t>Any change made to the "revision-date" or "revision-or-derived" substatements
```
[the remark this answers is not in the document]
Comment (jsterne) [reply yang-module-versioning/draft-ietf-netmod-yang-module-versioning.xml:206]: done in local draft - will be part of next checkin in this PR.

File: yang-module-versioning/draft-ietf-netmod-yang-module-versioning.xml (hunk 8 above)
Comment (jsterne) [reply yang-module-versioning/draft-ietf-netmod-yang-module-versioning.xml:578]: done in local draft - will be part of next checkin in this PR.

File: yang-module-versioning/draft-ietf-netmod-yang-module-versioning.xml (hunk 5 above)
Comment (jsterne) [reply yang-module-versioning/draft-ietf-netmod-yang-module-versioning.xml:590]: In the last meeting we proposed that we wouldn't preclude this, but would say they must be removed before publishing a released version. Done in local draft - will be part of next checkin in this PR.

**Comment by jsterne** [comment]
Merging in Master as discussed in previous weekly calls. Further changes will be dealt with as part of general reviews and other items tracked as Issues.
